### ISSUE TITLE
How to use `earthly` during development?

### ISSUE THREAD

**Issue by mpdude** [earthly/earthly#1517]
👋🏼 Hello everyone, 

this is my first issue here in Earthly.

I am currently looking into how our development and build workflow could benefit from switching to a container-based approach. I think I get the basic idea of Earthly and I see how it helps making builds repeatable.

But there is one question with regard to development that I find unanswered:

My goal is to use dockerized versions of all the tools I need during development. That hopefully makes it easier for people new to a project, because there is no need for complex VM setups and installing tools. Also, it should make it easier to use different versions of a tool in different projects.

So, let's assume I am working on a Node-based project and I don't have `yarn` installed locally. Now I would like to `yarn add some/dependency`. 

Is there some way Earthly can help me with that? I mean, instead of writing

`docker run --rm -it -v $(pwd):$(pwd) --workdir ...blah... user...ssh-socket... node:16 yarn add foo`

... I'd like to do something like `earthly +yarn add foo` (or whatever syntax).

That should do something like running the `node:16` container, a mounting my project directory into it.

Similarly, when I'd like to spin up a webserver for some experiments, I'd like to run `docker run ... node:16 yarn exec ts-node index.ts` or so, exposing a port and keeping it running until I either `Ctrl-C` (or even keep it running in the background).

Is that possible with Earthly?

**Comment by vladaionescu** [comment]
Hi @mpdude - it's not possible quite the way you describe it, however, you can use the command `LOCALLY` to execute commands on the host system, outside of containers.

Alternatively, you could create targets for adding dependencies: `earthly +add-yarn-dep --dep=some/dependency` and the target might output the `package.json` to replace the one in your current directory. This would be closer to what you have in mind. Something like this:

```
add-yarn-dep:
   FROM node:16
   WORKDIR /project
   COPY package*.json ./
   ARG dep
   RUN yarn add "$dep"
   SAVE ARTIFACT package*.json AS LOCAL ./
```

If you wanted to drop into a shell in the container you can also do `RUN --interactive node`.

**Comment by mpdude** [comment]
Thank you @vladaionescu for the quick answer!

To stick with the `yarn` example, there is a lot of subcommands (`install`, `add`, `upgrade`, ...) that one might want to run during development, and creating targets for each of them is cumbersome. Also, passing additional or optional arguments to these commands seems to be very involved when you have to use build arguments for all of them.

Is my impression right that Earthly is mostly about running and orchestrating Docker _build_ processes, with strong support for dependencies between those and and "artifacts" being a way to copy files out of resulting images into new build processes and/or to the local filesystem?

I have seen the JS example at https://github.com/earthly/earthly/tree/main/examples/tutorial/js and understand how that comes to a final Docker image that runs my application in a Node-based HTTP server.

But how would I use Earthly in that example when I want to work on the application locally? I probably need the vendors (`node_modules`) to be in my local filesystem path as well, and I need to run the container with the server with a volume mount... 

Is `LOCALLY` the answer to that, effectively putting commands into the `Earthfile` that otherwise would be in a shell script of some kind?

**Comment by vladaionescu** [comment]
Yeah, currently `LOCALLY` is the main answer to that. Earthly doesn't yet replace typical local workflows. We may have more features around that in the future, though.

**Comment by PyB1l** [comment]
Do you have any updates on this?

**Comment by fdietze** [comment]
I suggest to use https://www.jetify.com/devbox for your use-case.
Devbox can also be used to install the exact same versions of software into a container. That way development and production deployment are very close and everyone is using the same tools.
